### PR TITLE
expansion into orthogonal series on the reference element and step-27 bugfix/update

### PR DESCRIPTION
**Pull request by davydden**
added a couple of classes to do expansion of FE field on a reference element into orthogonal bases: Fourier and Legendre.

**Step-27**:
Step-27 appears to have a bug in angular frequencies (k vectors). The multiplier should be `2 \pi` instead of `pi`. Otherwise exponential functions are not hermitian orthogonal on `[0,1]` (Maxima code):
```
a:0;
b:1;
define(Phi(n), exp(-2*%i*%pi*n*x/(b-a)));
integrate(Phi(0)*conjugate(Phi(0)),x,a,b); /* 1 */
integrate(Phi(1)*conjugate(Phi(1)),x,a,b); /* 1 */
integrate(Phi(1)*conjugate(Phi(2)),x,a,b); /* 0 */
integrate(Phi(2)*conjugate(Phi(2)),x,a,b); /* 1 */
```
and in 2D
```
define(Phi2D(n,m), exp(-2*%i*%pi*(n*x+m*y)));
integrate(integrate(Phi2D(0,0)*conjugate(Phi2D(0,0)),x,0,1),y,0,1); /* 1 */
integrate(integrate(Phi2D(1,2)*conjugate(Phi2D(1,2)),x,0,1),y,0,1); /* 1 */
integrate(integrate(Phi2D(1,2)*conjugate(Phi2D(3,4)),x,0,1),y,0,1); /* 0 */
integrate(integrate(Phi2D(1,2)*conjugate(Phi2D(2,1)),x,0,1),y,0,1); /* 0 */
integrate(integrate(Phi2D(3,3)*conjugate(Phi2D(3,3)),x,0,1),y,0,1); /* 1 */;
```
Also see MWE in Maxima in `tests/fe/fe_series_01.cc` which illustrates that the implementation in this PR  is correct.

**TODO**:
- [x] 2d/3d tests
- [x] document Legendre
- [x] entry in changes.h
- [x] mention/use Fourier class in step-27
- [x] update results section in step-27
- [x] the updated step27 with `FESeries` classes and methods gives the same refinement path as the original implementation with corrected k vectors. Add it to the unit test.

**Inconviniencies**:
For transformation classes `hp::FECollection` and `hp::QCollection` have to be alive at the time of `calculate()` call. Since there are no `clone()`, so far i don't see an easy way out of this problem. *update:* at least currently this case is guarded by using `SmartPointers`.

### CODE REVIEW

File: include/deal.II/fe/fe_series.h [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+
+    /**
+     * Angular frequencies \f$ 2 \pi {\bf k} \f$ .
+     */
+    Table<dim, Tensor<1,dim> > k_vectors;
+
+    /**
+     * Transformation matrices for each FiniteElement.
+     */
+    std::vector<std::shared_ptr< FullMatrix<std::complex<double> > > > fourier_transform_matrices;
```
Comment (drwells) [inline include/deal.II/fe/fe_series.h]: I do not believe that the `shared_ptr` is needed. I think that everything you do below with `ensure_existence` would work where you check if the size of the matrix is zero instead.

Comment (davydden) [inline include/deal.II/fe/fe_series.h]: agree. will switch to that.

**Comment by davydden** [comment]
Here is a comparison of old-vs-new `estimate_smoothness`:
```
  template <int dim>
  void
  LaplaceProblem<dim>::
  estimate_smoothness (Vector<float> &smoothness_indicators)
  {
#ifdef OLD
    const unsigned int N = max_degree;

    std::vector<Tensor<1,dim> > k_vectors;
    std::vector<unsigned int>   k_vectors_magnitude;
    switch (dim)
      {
      case 2:
      {
        for (unsigned int i=0; i<N; ++i)
          for (unsigned int j=0; j<N; ++j)
            if (!((i==0) && (j==0))
                &&
                (i*i + j*j < N*N))
              {
                k_vectors.push_back (Point<dim>(2.*numbers::PI * i,
                                                2.*numbers::PI * j));
                k_vectors_magnitude.push_back (i*i+j*j);
              }

        break;
      }

      case 3:
      {
        for (unsigned int i=0; i<N; ++i)
          for (unsigned int j=0; j<N; ++j)
            for (unsigned int k=0; k<N; ++k)
              if (!((i==0) && (j==0) && (k==0))
                  &&
                  (i*i + j*j + k*k < N*N))
                {
                  k_vectors.push_back (Point<dim>(2.*numbers::PI * i,
                                                  2.*numbers::PI * j,
                                                  2.*numbers::PI * k));
                  k_vectors_magnitude.push_back (i*i+j*j+k*k);
                }

        break;
      }

      default:
        Assert (false, ExcNotImplemented());
      }

    const unsigned n_fourier_modes = k_vectors.size();
    std::vector<double> ln_k (n_fourier_modes);
    for (unsigned int i=0; i<n_fourier_modes; ++i)
      ln_k[i] = std::log (k_vectors[i].norm());


    std::vector<Table<2,std::complex<double> > >
    fourier_transform_matrices (fe_collection.size());

    QGauss<1>      base_quadrature (2);
    QIterated<dim> quadrature (base_quadrature, N);


    for (unsigned int fe=0; fe<fe_collection.size(); ++fe)
      {
        fourier_transform_matrices[fe].reinit (n_fourier_modes,
                                               fe_collection[fe].dofs_per_cell);

        for (unsigned int k=0; k<n_fourier_modes; ++k)
          for (unsigned int j=0; j<fe_collection[fe].dofs_per_cell; ++j)
            {
              std::complex<double> sum = 0;
              for (unsigned int q=0; q<quadrature.size(); ++q)
                {
                  const Point<dim> x_q = quadrature.point(q);
                  sum += std::exp(std::complex<double>(0,1) *
                                  (k_vectors[k] * x_q)) *
                         fe_collection[fe].shape_value(j,x_q) *
                         quadrature.weight(q);
                }
              fourier_transform_matrices[fe](k,j)
                = sum;
            }
      }

    std::vector<std::complex<double> > fourier_coefficients (n_fourier_modes);
    Vector<double>                     local_dof_values;

    typename hp::DoFHandler<dim>::active_cell_iterator
    cell = dof_handler.begin_active(),
    endc = dof_handler.end();
    for (; cell!=endc; ++cell)
      {
        local_dof_values.reinit (cell->get_fe().dofs_per_cell);
        cell->get_dof_values (solution, local_dof_values);

        for (unsigned int f=0; f<n_fourier_modes; ++f)
          {
            fourier_coefficients[f] = 0;

            for (unsigned int i=0; i<cell->get_fe().dofs_per_cell; ++i)
              fourier_coefficients[f] +=
                fourier_transform_matrices[cell->active_fe_index()](f,i)
                *
                local_dof_values(i);
          }

        std::map<unsigned int, double> k_to_max_U_map;
        for (unsigned int f=0; f<n_fourier_modes; ++f)
          if ((k_to_max_U_map.find (k_vectors_magnitude[f]) ==
               k_to_max_U_map.end())
              ||
              (k_to_max_U_map[k_vectors_magnitude[f]] <
               std::abs (fourier_coefficients[f])))
            k_to_max_U_map[k_vectors_magnitude[f]]
              = std::abs (fourier_coefficients[f]);

        double  sum_1           = 0,
                sum_ln_k        = 0,
                sum_ln_k_square = 0,
                sum_ln_U        = 0,
                sum_ln_U_ln_k   = 0;
        for (unsigned int f=0; f<n_fourier_modes; ++f)
          if (k_to_max_U_map[k_vectors_magnitude[f]] ==
              std::abs (fourier_coefficients[f]))
            {
              sum_1 += 1;
              sum_ln_k += ln_k[f];
              sum_ln_k_square += ln_k[f]*ln_k[f];
              sum_ln_U += std::log (std::abs (fourier_coefficients[f]));
              sum_ln_U_ln_k += std::log (std::abs (fourier_coefficients[f])) *
                               ln_k[f];
            }

        const double mu
          = (1./(sum_1*sum_ln_k_square - sum_ln_k*sum_ln_k)
             *
             (sum_ln_k*sum_ln_U - sum_1*sum_ln_U_ln_k));

        smoothness_indicators(cell->active_cell_index()) = mu - 1.*dim/2;
      }
#else
    Vector<double>                     local_dof_values;

    typename hp::DoFHandler<dim>::active_cell_iterator
    cell = dof_handler.begin_active(),
    endc = dof_handler.end();
    for (; cell!=endc; ++cell)
      {
        local_dof_values.reinit (cell->get_fe().dofs_per_cell);
        cell->get_dof_values (solution, local_dof_values);

        fourier->calculate(local_dof_values,
                           cell->active_fe_index(),
                           fourier_coefficients);

        std::pair<std::vector<unsigned int>, std::vector<double> > res =
          FESeries::process_coefficients<dim>(fourier_coefficients,
                                              predicate_ind<dim>,
                                              VectorTools::Linfty_norm);

        Assert (res.first.size() == res.second.size(),
                ExcInternalError());

        if (ln_k.size() == 0)
          {
            ln_k.resize(res.first.size(),0);
            for (unsigned int f = 0; f < ln_k.size(); f++)
              ln_k[f] = std::log (2.0*numbers::PI*std::sqrt(1.*res.first[f]));
          }

        for (unsigned int f = 0; f < res.second.size(); f++)
          res.second[f] = std::log(res.second[f]);

        std::pair<double,double> fit = FESeries::linear_regression(ln_k,res.second);
        smoothness_indicators(cell->active_cell_index()) = -fit.first - 1.*dim/2;
      }
#endif
  }
```

**Comment by davydden** [comment]
here is a smoothness estimator for the new algorithm (with `2\Pi`). Looking at the solution (https://dealii.org/8.4.1/doxygen/deal.II/step_27.html#Results), i think it is closer to the reality as the previous version.
step-27 smoothness-0

**Comment by davydden** [comment]
I did everything i wanted to do here, it's ready for review. I know that you Wolfgang @bangerth is busy these days, but since you are the only author of `step-27` and the main author of the `hp paper`, i guess we won't be able to proceeded without your feedback on this PR.

File: doc/news/changes.h [hunk 2]
```diff
@@ -120,6 +120,16 @@ inconvenience this causes.
 <h3>General</h3>
 
 <ol>
+ <li> New: add new classes to expand scalar finite element solution into
```
Comment (kronbichler) [inline doc/news/changes.h]: Capitalize 'Add'

File: doc/news/changes.h [hunk 3]
```diff
@@ -120,6 +120,16 @@ inconvenience this causes.
...
+ orthogonal bases FESeries::Fourier and FESeries::Legendre. Also
+ provide an auxiliary functions to calculate norms of subsets of expansion
```
Comment (kronbichler) [inline doc/news/changes.h]: an auxiliary functions -> auxiliary functions

File: doc/news/changes.h [hunk 4]
```diff
@@ -120,6 +120,16 @@ inconvenience this causes.
...
+ coefficients FESeries::process_coefficients and linear regression
+ FESeries::linear_regression. Update step-27 to use this namespace to drive
+ hp-adaptive FEM solution process.
```
Comment (kronbichler) [inline doc/news/changes.h]: insert a 'the' before hp-adaptive

File: examples/step-27/doc/intro.dox [hunk 5]
```diff
@@ -312,27 +312,30 @@ Assuming that the cell $K$ is not degenerate, i.e. that the mapping from the
 unit cell to cell $K$ is sufficiently regular, above condition is of course
 equivalent to
 @f[
-	\int_{\hat K} |\nabla^s \hat u(\hat{\bf x})|^2 \; d\hat{\bf x} < \infty
+	\int_{\hat K} |\nabla^s \hat u(\hat{\bf x})|^2 \; d\hat{\bf x} < \infty\,,
 @f]
-where $\hat u(\hat{\bf x})$ is the function $u({\bf x})$ mapped back onto the unit cell
+where $\hat u(\hat{\bf x})$ is the function $u({\bf x})$ mapped back onto the reference unit cell
```
Comment (kronbichler) [inline examples/step-27/doc/intro.dox]: Is there a reason you say 'reference unit cell' now? Isn't it either the reference cell or the unit cell?

Comment (davydden) [inline examples/step-27/doc/intro.dox]: they are the same, but i agree that one should use either one or another.

File: examples/step-27/doc/intro.dox [hunk 6]
```diff
@@ -312,27 +312,30 @@ Assuming that the cell $K$ is not degenerate, i.e. that the mapping from the
...
-	\hat U_{\bf k}
-	= \frac 1{(2\pi)^{d/2}} \int_{\hat K} e^{i {\bf k}\cdot \hat{\bf x}} \hat u(\hat{\bf x}) d\hat{\bf x}
+	\hat u(\hat{\bf x})
+	= \sum_{\bf k} \hat U_{\bf k}\,e^{-i {\bf k}\cdot \hat{\bf x}},
 @f]
 with Fourier vectors ${\bf k}=(k_x,k_y)$ in 2d, ${\bf k}=(k_x,k_y,k_z)$
-in 3d, etc, and $k_x,k_y,k_z=0,\pi,2\pi,3\pi,\ldots$. If we re-compose $\hat u$
-from $\hat U$ using the formula
+in 3d, etc, and $k_x,k_y,k_z=0,2\pi,4\pi,\ldots$. The coefficients of expansion
+$\hat U_{\bf k}$ can be obtained using $L^2$-orthogonality of exponential basis
```
Comment (kronbichler) [inline examples/step-27/doc/intro.dox]: of exponential basis -> of the exponential basis

File: examples/step-27/step-27.cc [hunk 7]
```diff
@@ -164,6 +172,29 @@ namespace Step27
...
+  }
+
+  template <typename T>
+  void resize(Table<3,T> &coeff, const unsigned int N)
+  {
+    TableIndices<3> size;
+    for (unsigned int d=0; d<3; d++)
+      size[d] = N;
+    coeff.reinit(size);
+  }
```
Comment (kronbichler) [inline examples/step-27/step-27.cc]: Can't you use the templated `TableIndices<dim>` class to cover all cases at once? Then you can avoid using these specializations...

File: examples/step-27/step-27.cc [hunk 8]
```diff
@@ -176,6 +207,48 @@ namespace Step27
...
+    // program. The FESeries::Fourier class' constructor first parameter $N$ defines
+    // the number of coefficients in 1D with the total number of coefficients
+    // being $N^{dim}$. Although we will not use coefficients corresponding to
+    // $\sqrt{i^2+j^2}> N$ and $i+j==0$, the overhead of their calculation is minimal.
+    // The transformation matrices for each FiniteElement will be calculated only
+    // once the first time they are required in the course of hp-adaptive
+    // refinement. Because we work on the unit cell, we can do all this work without a
+    // mapping from reference to real cell and consequently can precalculate
+    // these matrices. The calculation of expansion
+    // coefficients for a particular set of local degrees-of-freedom on a given
```
Comment (kronbichler) [inline examples/step-27/step-27.cc]: Please write 'degrees of freedom' rather than 'degrees-of-freedom'

File: examples/step-27/step-27.cc [hunk 9]
```diff
@@ -612,141 +685,63 @@ namespace Step27
...
   // @sect4{LaplaceProblem::estimate_smoothness}
 
+  // As described in the introduction, we will need to take the maximum
+  // absolute value of fourier coefficients which correspond to $k$-vector
+  // $|{\bf k}|= const$. To filter the coefficients Table we
+  // will use the FESeries::process_coefficients() which requires a predicate
+  // to be specified. The predicate should operate on TableIndices and return
+  // a pair of <code>bool</code> and <code>unsigned int</code>. The latter
+  // is the value of the map from TableIndicies to unsigned int.  It is
+  // be used to define subsets coefficients from which we search for the one
```
Comment (kronbichler) [inline examples/step-27/step-27.cc]: remove 'be', say 'subsets of coefficients' rather than 'subsets coefficients' (?)

File: examples/step-27/step-27.cc [hunk 10]
```diff
@@ -612,141 +685,63 @@ namespace Step27
...
+  // with highest absolute value, i.e. $l^\infty$-norm. Whereas <code>bool</code>
+  // parameter defiens which indices should be used in processing. In the
```
Comment (kronbichler) [inline examples/step-27/step-27.cc]: Make the sentence over the last two lines start somewhere along the lines of "The `<code>bool</code>` parameter defines which..."

File: examples/step-27/step-27.cc [hunk 11]
```diff
@@ -612,141 +685,63 @@ namespace Step27
...
+  template<>
+  std::pair<bool,unsigned int>
+  predicate_ind<3>(const TableIndices<3> &ind)
+  {
+    const unsigned int v = ind[0]*ind[0]+ind[1]*ind[1]+ind[2]*ind[2];
+    if (v>0 && v < 5*5*5)
+      return std::make_pair(true,v);
+    else
+      return std::make_pair(false,v);
+  }
```
Comment (kronbichler) [inline examples/step-27/step-27.cc]: Isn't there a dimension-independent way for doing this? I can't see anything special do each function, except that you have different upper bounds (is this the maximal polynomial degree)?

Comment (davydden) [inline examples/step-27/step-27.cc]: correct. Perhaps one should wrap it into the class function and to `std::bind()` instead.

File: include/deal.II/fe/fe_series.h [hunk 12]
```diff
@@ -0,0 +1,428 @@
...
+DEAL_II_NAMESPACE_OPEN
+
+
+/*!@addtogroup feall */
+/*@{*/
+
+
+/**
+ * This namespace offers functions to calculate expansion series of the
+ * solution on the reference element. Coefficient of expansion are often used
```
Comment (kronbichler) [inline include/deal.II/fe/fe_series.h]: Coefficient -> Coefficients

File: include/deal.II/fe/fe_series.h [hunk 13]
```diff
@@ -0,0 +1,428 @@
...
+   * this class. Note that \f$ u({\bf x}) = \sum_i u_i N_i({\bf x})\f$,
+   * where \f$ N_i({\bf x}) \f$ are real-valued FiniteElement shape functions.
+   * Consequently \f$ c_{\bf k} \equiv c_{-\bf k}^\ast \f$ and
+   * we only need to compute \f$ c_{\bf k} \f$ for positive indices
+   * \f$ \bf k \f$ .
+   *
+   * @author Denis Davydov, 2016.
+   */
+  template <int dim>
+  class Fourier : public Subscriptor
```
Comment (kronbichler) [inline include/deal.II/fe/fe_series.h]: Is there a reason you derive this class (and Legendre further down) from Subscriptor, i.e., is there somewhere where a SmartPointer to this class is stored? (On the other hand, since this class uses a quadrature collection which we sometimes copy, sometimes not, I wonder whether the fe_collection and q_collection arguments should rather be `SmartPointer<...>` to make sure there is no dangling reference.

File: include/deal.II/fe/fe_series.h [hunk 14]
```diff
@@ -0,0 +1,428 @@
...
+    const unsigned int &pred_value = pred_pair.second;
+    const T &coeff_value = coefficients(ind);
+
+    if (pred_to_values.find (pred_value) == pred_to_values.end())
+      {
+        pred_to_values[pred_value] = std::vector<T>(1,coeff_value);
+      }
+    else
+      {
+        pred_to_values[pred_value].push_back(coeff_value);
```
Comment (kronbichler) [inline include/deal.II/fe/fe_series.h]: Isn't a `map` returning a reference to a newly created vector also in the case you don't find it, so `push_back` would also work in that case? (Please note I don't recall how exactly those things are processed, so I might be wrong.)

Comment (davydden) [inline include/deal.II/fe/fe_series.h]: if an entry does not exist, `pred_to_values[pred_value]` will create it, otherwise it returns the existing entry. As far as I can tell, the code above should be ok.

File: include/deal.II/fe/fe_series.h [hunk 15]
```diff
@@ -0,0 +1,428 @@
...
+                const std_cxx11::function<std::pair<bool,unsigned int>(const TableIndices<3> &)> &predicate,
+                std::map<unsigned int, std::vector<T> > &pred_to_values)
+  {
+    for (unsigned int i = 0; i < coefficients.size(0); i++)
+      for (unsigned int j = 0; j < coefficients.size(1); j++)
+        for (unsigned int k = 0; k < coefficients.size(2); k++)
+          {
+            const TableIndices<3> ind(i,j,k);
+            fill_map_index(coefficients,ind,predicate,pred_to_values);
+          }
```
Comment (kronbichler) [inline include/deal.II/fe/fe_series.h]: Similar to above, I wonder whether you could skip the specializations if you fill a `TableIndices<dim>` object with `operator[]`. Of course, where have the complication that the number of loops depends on the dimension. When I have these kind of loops, I typically just run outer loops from `0` to `1`. I'm not insisting in a change here, though...

File: include/deal.II/fe/fe_series.h (hunk 13 above)
Comment (davydden) [inline include/deal.II/fe/fe_series.h]: No, i don't store `SmartPointers`, i just thought it is a convention to derive `deal.ii` classes from `Subscriptor`. 

As for the fe_collection and q_collection, i agree. This is the current `inconvenience` i mention in the first post. I will try reworking the code with `SmartPointers`...

**Comment by kronbichler** [comment]
To me, this looks really good. Thanks for the extensive set of tests and the code. step-27 looks much better in my opinion. (Besides @bangerth I'm probably the person who knows most of step-27 since I re-worked the part for sparsity pattern generation, constraints, and so on and used it for performance measurements in `ConstraintMatrix::distribute_local_to_global`. I actually also worked on some hp code too, even though my own work was in the context of MatrixFree). I guess we leave this open a few more days in case someone has suggestions.

**Comment by kronbichler** [comment]
And of course, you should squash into a few commits when we are done...

File: include/deal.II/fe/fe_series.h (hunk 15 above)
Comment (davydden) [inline include/deal.II/fe/fe_series.h]: i think the problem here is the nested loops which depend on the dimension. It would be good to have a `.begin()` `.end()` iterator for `Table` with `it->first` being `TableIndices` and `it->second` being coefficient values, but this is outside of the scope of this PR.

File: include/deal.II/fe/fe_series.h (hunk 14 above)
Comment (kronbichler) [inline include/deal.II/fe/fe_series.h]: Yes, your code should be OK, I was just wondering if you could get rid of the if/else statement here...

File: include/deal.II/fe/fe_series.h (hunk 13 above)
Comment (kronbichler) [inline include/deal.II/fe/fe_series.h]: Now I understand the first post better, thanks. I think that SmartPointers should be exactly the thing you're looking for. Both the fe collection and q collection should support that.

File: include/deal.II/fe/fe_series.h (hunk 15 above)
Comment (kronbichler) [inline include/deal.II/fe/fe_series.h]: I agree. And I guess we are fine with this code. As I realized while writing the original message, you don't win very much here...

File: include/deal.II/fe/fe_series.h (hunk 14 above)
Comment (davydden) [inline include/deal.II/fe/fe_series.h]: oh, i see, let me check...

**Comment by davydden** [comment]
@kronbichler thanks for reviewing. It is always good to have a second pair of eyes :+1: :smile:.

**Comment by davydden** [comment]
p.s. at some point we will need to update the figures on the website (https://www.dealii.org/images/steps/developer/step-27.mesh-0.png and alike), i have them generated but they appear to be outside of GitHub repo.

**Comment by bangerth** [comment]
Neat -- so there was just a plain and simple bug!

I certainly won't get to it for several days due to travel, but will be happy to look at it after that unless @kronbichler thinks it's ready even without my eyes.

**Comment by kronbichler** [comment]
@davydden One thing I was curious about when I saw the picture: Does your improved algorithm fix the issue with the 'bad smoothness' indicated at cells with hanging nodes observed in step-27. In other words, how does the smoothness in the final picture of step-27 look like?

**Comment by davydden** [comment]
@bangerth i think it's fine to wait a couple of days when you will have a chance to give it a quick look :smile:

**Comment by davydden** [comment]
@kronbichler the final smoothness is below. You can see that roughly at the position of the peak the smoothness is low (as expected). But there are also some strange indicators for some elements. I think the issue is still there.
step-27 smoothness-5

**Comment by kronbichler** [comment]
Thanks for sharing the picture. Overall, the mesh looks much more reasonable now. With regard to the smoothness, I agree that the behavior is the same. The explanation that the solution representation is similar in nearby elements due to the constraints but appears "smoothers" on the refined side makes sense after all.

**Comment by davydden** [comment]
For completeness, here is the plot of final degrees
step-27 fe_degree-5

**Comment by davydden** [comment]
@bangerth ping (step-27 update/bugfix).

**Comment by davydden** [comment]
p.s. merge conflicts are with `changes` file, will fix it right before merging as it appears again and again.

**Comment by davydden** [comment]
@bangerth @kronbichler : i would like to finish with this PR until the middle of the next week as i will go for long holidays. If there are further comments, please let me know. I am fine to wait as well, but I would prefer to roughly know when we'll wrap it up.

File: doc/news/changes.h [hunk 16]
```diff
@@ -120,13 +120,25 @@ inconvenience this causes.
 <h3>General</h3>
 
 <ol>
+
+ <li> New: Add new classes to expand scalar finite element solution into
+ orthogonal bases FESeries::Fourier and FESeries::Legendre. Also
```
Comment (kronbichler) [inline doc/news/changes.h]: I would add two articles to the first sentence: "Add new classes to expand a scalar finite element solution into the orthogonal bases FESeries::Fourier and FESeries::Legendre"

File: doc/news/changes.h [hunk 17]
```diff
@@ -120,13 +120,25 @@ inconvenience this causes.
...
+ </li>
+
  <li> New: Add VectorTools::compute_global_error that computes global
  errors from cellwise errors obtained by VectorTools::integrate_difference()
  and do MPI collectives if necessary.
  <br>
  (Timo Heister, 2016/05/15)
  </li>
 
+
```
Comment (kronbichler) [inline doc/news/changes.h]: Remove this

**Comment by kronbichler** [comment]
From my point of view, this looks ready for merging. :+1: to the extensive set of tests.

**Comment by bangerth** [comment]
I wanted to look at this, and it even made it to position 2 on my TODO list, but I haven't gotten to it :-( I'm at the ASPECT hackathon till the end of next week, so nothing much is going to happen till then. If @kronbichler is happy with it, I'll gladly let him merge. I might still look at the patch later on, and if I find something I'll fix it myself (as punishment for being slow).

**Comment by davydden** [comment]
@bangerth we need to update images (meshes/solution) on the website, which are referenced in this tutorial. I have them ready, whom should I send them?

**Comment by kronbichler** [comment]
@davydden Do you have access to the deal.II homepage svn repository? (I have, in case you do not.)

The files are `https://svnhp.dealii.org/trunk/homepage/images/steps/developer/step-27.*`

Another point regarding this PR: Could you please fix the C++11-related test failures here:
http://cdash.kyomu.43-1.org/viewTest.php?onlyfailed&buildid=1130

**Comment by davydden** [comment]
@kronbichler i don't think i have an access, here are all the images, could you please upload them? 
[Archive.zip](https://github.com/dealii/dealii/files/335029/Archive.zip)

w.r.t. the failing test, i don't understand why it fails. I though that `std_cxx11::shared_ptr` should use `std` when `c++11` is available and `boost` otherwise. In other words, it should compile in both scenarios.

**Comment by Rombur** [comment]
@davydden you need to  write ```std_cxx11::shared_ptr<FESeries::Fourier<dim> >``` instead of ```std_cxx11::shared_ptr<FESeries::Fourier<dim>>```

**Comment by davydden** [comment]
oh, i was confused by `warning` there, will do...
